### ISSUE TITLE
Mapping WoT event models in Scripting API

### ISSUE THREAD

**Issue by zolkis** [w3c/wot-scripting-api#68]
This issue is to track WG discussion on WoT event models (in TD and Bindings TF), and how to represent them in the WoT Scripting API.

**Comment by zolkis** [comment]
As input, summarizing a few points from [TD Events](https://lists.w3.org/Archives/Member/member-wot-wg/2017Aug/att-0018/td-events.pdf) slide. It is work in progress.

**TD Event Model**
- Event interaction provides output data with zero or more property values, e.g. state change of a property
- Event interaction verbs: subscribe, unsubscribe
- Protocol binding template describes how to obtain events and parse property values.

**Event examples**
- Button states: pressed, released
- Button events: pressed, shortpress, longpress, doublepress, held, released
- Is there an event resource / interaction instance per event type, or a single source with multiple types?

**Event queueing**
- Events may be optionally stored until processed
- There could be a queue of generated events
- Ideally, there is a queue per subscriber, but there could be a useful shared queue
- Subscribing could create a queue, managed by the subscriber
- Pubsub protocols might provide queueing
- Link Binding can point to queues
- How do we represent this in a TD?

**Conditional Notification**
- Use case: _conditional_ notification on property value change
- Time + value based filtering
    - notify on change of at least <delta>
    - notify min, max interval
    - notify in/out of signal band
- How do we represent this in a TD?

**Event Protocol Bindings**
- CoAP observe 
- MQTT, other Pub/Sub
- WebSockets
- HTTP, HTTP Long poll
- EventSource
- Link Bindings to events: hyperlink pointing to event sink/destination resource, i.e. the server side of a "web callback"

**Questions and issues**
- Groups of events to subscribe to, notion of grouping
- Event abstraction + metadata independent of protocol, e.g. high/low level of the water tank
- W3C related work on events: DOM Event, MMI/HMI, automotive (ex. conditional notifications in [VSS server spec](https://www.w3.org/TR/vehicle-information-service/))
- separate protocol and event metadata
- protocol may restrict the options available on a particular event, how should we restrict the interaction model? URI schemes, also see capability/ontology model
- practical examples of what is currently out there e.g.  CoAP, MQTT, WS
- pattern to describe: how to subscribe, create/modify/delete subscription resource, filter criteria, related properties, way to get notified: link, notification URI (callback target), etc 
- Different types of events: predefined by the device, application-defined, e.g. client can subscribe and customize
- Also a topic for binding templates, scripting API, and vocabulary
- Property values with timestamps
- Conditional spec is application specific, also can be a capability/ontology element that can describe the different capabilities

Scripting API may use Event (DOM or Node) or Observable. Currently betting on Observable.

**Comment by zolkis** [comment]
Possible requirements for scripting API:
- Data carried by the event is `any`: described by a TD fragment; could be number, string, date, timestamp, object, array etc
- Event queuing on which side? 
    - Sender side would be new, and would require the protocol to ack handling the event. Delayed events may be an issue in many applications.
    - Receiver side queuing is implementation specific.
- conditional notifications, e.g. a filter (dictionary or boolean function) passed when subscribing to an event
- event grouping, e.g. label an array of event names that acts like an aggregated event; notifications will carry an array of events
    - a group event is fired when any event fires?
    - or is there a low-pass filter for firing, e.g. "send all events that fired in the past minute" --> conditional notifications supported on event groups, too

**Comment by zolkis** [comment]
What could be supported in Scripting API:
- Events could be modeled by [DOM Events](https://dom.spec.whatwg.org/#events), [Node.js Events](https://nodejs.org/api/events.html) or [Observables](https://github.com/tc39/proposal-observable). Suggesting the latter.
- Support for conditional notification involves passing an argument when subscribing: e.g. a dictionary or a boolean function.
- Support for event groups involves passing the list of group names this event is part of, when subscribing.

Supporting all these could be achieved by modifying the [observe()](https://w3c.github.io/wot-scripting-api/#the-observe-method) signature as follows:

```javascript
partial interface ConsumedThing {
    Observable observe(DOMString name, optional ObserveOptions options);
};

dictionary ObserveOptions {
    callback filter = boolean (optional any value);
    // property-value pairs for constraints, defined by the TD
};
```
Above is just an example for possible `ObserveOptions`. There, if `ObserveOptions.filter` is a `Function`, it is used for filtering and the event data is passed to it. If any properties are defined on `options`, they are used as further constraints on the corresponding properties on the data carried by the event. 

For instance, if a property named `temperature` has the value `20` on the `options`, the event is fired only when the same property has the same value on the event data. If filtering is more complex than just comparing values,  e.g. when value conversions or range comparisons are needed, the app should define the `filter` function. If the `filter` function is defined, and `options` also contains properties, the event is fired only if both the function returns `true` and all property values match a corresponding property on the event data.

**Comment by mkovatsc** [comment]
> Support for conditional notification 

Is the assumption that the filtering is done on the client-side? Overall, W3C WoT assumes that the devices are fixed and we cannot expect them to suddenly support, for instance, conditional notifications.

So far, there is no activity on this in the TD Task Force; I would see the priority also too low to deal with this now. For client-side filtering, I could also see a convenience library on top doing this -- the "damage" of sending a message/wasting energy is done already anyway. I am pointing this out, because I think we should focus on a stable core first before starting new features.

> Event interaction provides output data with zero or more property values, e.g. state change of a property

As written elsewhere, we need to distinguish between Change-of-Value notifications and events. Many IoT platforms and people get this wrong and do not understand the implications. See #73:

As of now, the TD Interaction Model foresees

* Observable Properties, which can push change-of-value notifications (which only need eventual consistency, so intermediate notifications may be lost or even filtered to reduce the load)
* Events, which define sources of individual event instances. Clients can subscribe to sources and each event instance must be transferred from the source to the client.

Please update the summary here.

**Comment by zolkis** [comment]
> Is the assumption that the filtering is done on the client-side? 

There is no assumption, just stating the material of the TD TF.
Since the API is client-side, my first guess would be client-side filtering, unless a server doesn't explicitly support it (and in that case it needs to be formalized in its TD).

> Overall, W3C WoT assumes that the devices are fixed 

Then why were you taking the example of adding a property during running a Thing in #76?

> So far, there is no activity on this in the TD Task Force; I would see the priority also too low to deal with this now. 

Very good, one less complication to solve. However, when time eventually comes for it, we at least have thought about it: `observe()` will just need an optional argument, so no future API break.

> As written elsewhere, we need to distinguish between Change-of-Value notifications and events. 

Please discuss that with the TD TF, they have written the text you are arguing with. And come back with the conclusion :).

> Events, which define sources of individual event instances. Clients can subscribe to sources and each event instance must be transferred from the source to the client.

Events can be lost as well. Clients usually deal with that. If WoT wants to _guarantee_ event delivery, it has to set up a fail-safe network infrastructure. Any examples on that?

> Please update the summary here.

Hmm, what summary?

**Comment by mkovatsc** [comment]
summary = your extensive lists above.

Another try: when Change-of-Value notifications are dropped, it is not an error case, it can even be desirable to avoid congestion. When Events are dropped, it is an error case and people want to avoid given the cost to do so. Please combine this piece of information with the other explanations I tried to give you on this topic.

**Comment by zolkis** [comment]
This is an interpretation. For some solutions dropped value change notifications are as much errors as dropped events, and for others losing samples is not an issue (a late sample is worse).

Also, a dropped event is not necessarily an error - it depends on its context, supposed to be defined by the TD.

Events are just a mechanism to deliver notifications. There are a couple of event systems out there and I don't see a generic trait you claim for them. It is another thing if the WoT WG has defined constraints on the terms Event and Notification - if that is the case, please provide a reference.
 
A TD is free to specify value changes by events.

Value changes also do have sources, some of them constrained by sample rate, jitter etc, some others are just pollable information. That behaviour should be modeled in the TD.

It is very confusing when you come up with conventions (or opinions) that I don't know whether have they been agreed in the WG, externally, or are they your opinions.

**Comment by mkovatsc** [comment]
It is one of the long-term discussions in the IRTF Thing-to-Thing Research Group, because it causes a lot of issues in IoT systems and for instance a lot of false expectations from CoAP Observe (e.g., in OMA LWM2M). Also is it native to exsiting systems such as BACnet.

**Comment by zolkis** [comment]
Addressed by #86.
